### ISSUE TITLE
Add Testing Platform Support

### ISSUE THREAD

**Issue by slang25** [stryker-mutator/stryker-net#3094]
Today Stryker has support for VSTest as the test runner, and this is what is commonly used today by various testing frameworks.

However Microsoft have built a new replacement call the [Microsoft Testing Platform](https://learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-platform-intro), which is being adopted by all of the major players (MSTest, NUnit, xUnit).

While many of these frameworks will support both runner engines for a good while, there are also new testing frameworks that are only support on the new testing platform (the excellent [TUnit](https://github.com/thomhurst/TUnit) for example).

It would be amazing if Stryker.NET add support for this as a new runner option.

**Comment by slang25** [comment]
I'll add that it may be possible to get indirect support today by using the [VSTest Bridge extension](https://learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-platform-extensions-vstest-bridge), which I haven't tested but assume would enable TUnit support today.

**Comment by rouke-broersma** [comment]
We are working on it, but it is a very large feature. We have to reimplement a lot of vstest features ourselves.

**Comment by richardwerkman** [comment]
I'm currently working on this. I'll take TUnit into account as well. Would be nice if we could support that.

**Comment by slang25** [comment]
Awesome, I didn't realise that this was being worked on, and how much work it is. I'm going to do some experimentation and see what I can get working today with TUnit,

**Comment by Lorilatschki** [comment]
First of all, thanks for the great tool, it really helps a lot qualifying our unit tests. We have currently moved to `xunit.v3` and the new `Microsoft Testing Platform`. We not directly using references to the MTP, we enable running via MSBuild properties.

```xml
<TestingPlatformDotnetTestSupport>true</TestingPlatformDotnetTestSupport>
<UseMicrosoftTestingPlatformRunner>true</UseMicrosoftTestingPlatformRunner>
<OutputType>Exe</OutputType>
```

Due to that, stryker does not run anymore when executed like.

```bash
dotnet-stryker --output .stryker/ --solution tests/stryker-test/src/all.sln --project Queue.csproj --test-project tests/stryker-test/src/Queue.Test/Queue.Test.csproj --mutation-level Advanced
```

```log
Version: 4.5.1
[13:10:36 INF] Analysis starting.
[13:10:36 INF] Analyzing 1 test project(s).
[13:10:51 INF] Found project /builds//tests/stryker-test/src/Queue/Queue.csproj to mutate.
[13:10:51 INF] Analysis complete.
[13:10:51 INF] Building test project /builds//tests/stryker-test/src/Queue.Test/Queue.Test.csproj (1/1)
[13:10:51 INF] Building project all.sln using dotnet build all.sln (directory /builds//tests/stryker-test/src.)
[13:10:57 INF] Number of tests found: 2 for project /builds//tests/stryker-test/src/Queue/Queue.csproj. Initial test run started.
[13:11:00 INF] 5 mutants created
[13:11:00 INF] Capture mutant coverage using 'CoverageBasedTest' mode.
[13:11:02 ERR] It looks like the test coverage capture failed. Disable coverage based optimisation.
Hint: by passing "--open-report or -o" the report will open automatically and 
update the report in real-time.
[13:11:02 INF] 2     mutants got status Ignored.      Reason: Removed by block already covered filter
[13:11:02 INF] 2     total mutants are skipped for the above mentioned reasons
[13:11:02 INF] 3     total mutants will be tested
Killed:   0
Survived: 3
Timeout:  0
```

Since you will add support for the `Microsoft Testing Platform`, is there currently any estimated timeframe, when we can expect stryker to run against the MTP?

**Comment by rouke-broersma** [comment]
Can't give you an estimation sorry, it's a lot of work.

**Comment by richardwerkman** [comment]
@Lorilatschki On your project, does stryker run succesful with `<TestingPlatformDotnetTestSupport>false</TestingPlatformDotnetTestSupport>`? If so, we could introduce a workaround in Stryker that disables this when building the project. Thay way you could use MsTestRunner on your project, but stryker will use VsTest to run the tests. 

Obviously we want to support running MsTest natively, but this will take some time to implement.

**Comment by richardwerkman** [comment]
I'm taking a new approach to implement this. Previously we made a POC that made Stryker the testhost. Basically stryker had to implement the MSTest.SDK and ran the test like this:
``` cs
            var testApplicationBuilder = await TestApplication.CreateBuilderAsync(args);
            // Register the testing framework
            testApplicationBuilder.AddMSTest(() => new[] { testAssembly });
            var factory = new CompositeExtensionFactory<StrykerExtension>(serviceProvider => new StrykerExtension());
            testApplicationBuilder.TestHost.AddDataConsumer(factory);
            var testApplication = await testApplicationBuilder.BuildAsync();

            await testApplication.RunAsync();
```
The problem is, that since Stryker is the testhost, the tests don't run in an external process. That means we can't use paralellization. The benefit is that we have access to the process so we can listen to all MsTest events.

Now I'm thinking to inject the above code into the test project, and setting the method as the entry point for the assembly. That way we override the default entrypoint. That way we have full control over the testrunner.

Outstanding issues:
- We will need to figure out how to add the custom extension and make it available in the stryker process. We need this to control the test execution and get the test results.
- We need to figure out how to keep the test process alive and run the tests multiple times for reduced overhead.
- We need to inject the code into the test project, for this we might need to build test projects with Roslyn. Currently we build test projects with `dotnet build`.

**Comment by Lorilatschki** [comment]
> [@Lorilatschki](https://github.com/Lorilatschki) On your project, does stryker run succesful with `<TestingPlatformDotnetTestSupport>false</TestingPlatformDotnetTestSupport>`? If so, we could introduce a workaround in Stryker that disables this when building the project. Thay way you could use MsTestRunner on your project, but stryker will use VsTest to run the tests.
> 
> Obviously we want to support running MsTest natively, but this will take some time to implement.

Unfortunately, setting the property `TestingPlatformDotnetTestSupport` to `false` leading to a failing stryker run. It states 

```
Number of tests found: 121 for project xxx.csproj. Initial test run started.
```

and than it got stucked (those tests are usually running ~5s).

**Comment by slang25** [comment]
Perhaps @nohwnd or @Evangelink could advise on a good hosting strategy, as MTP is quite flexible in terms of hosting options.

I guess there are multiple approaches available, including shelling out to the test executables directly (now they are self executing), or spinning up dedicated stryker subprocesses that act as the hosts.

To summarize the requirements of stryker we need (correct me if I'm wrong):
* Initial test run
  * With per-test code coverage collection
  * Capturing timings per-test
* Mutation test runs (_many times_)
  * As efficiently as possible
  * With concurrent test run execution
  * Changing an environment variable (active mutants) per-run
  * With custom timeout per-testrun
  * Capturing test results
  * Isolation between runs, i.e. some static state in one test run shouldn't bleed over to another

That last point might rule out a long lived server mode approach, which I think is something that MTP makes possible (assuming this because there is a HotRelead package for MTP).

**Comment by rouke-broersma** [comment]
Another big requirement is support for different target runtimes, since each test project can have a different runtime requirement. This almost immediately requires external test host orchestration. Before long we are rebuilding vstest 😅

**Comment by nohwnd** [comment]
I agree on the last point, if you get rid of per-process isolation there is no isolation left, and you cannot guarantee that mutations to static state won't bleed into next iterations. So that is probably the limitation that will need to be acknowledged and addressed.

There is hotreload plugin, details and implementation can be found here: https://github.com/microsoft/testfx/blob/main/src/Platform/Microsoft.Testing.Extensions.HotReload/HotReloadTestHostTestFrameworkInvoker.cs

In essence, it works by scheduling multiple runs inside of a test run, and that way keeping the process hosting the tests alive, because from the point of view of the process we did not finish running tests yet.

Similar thing could be done by stryker, but it does not guarantee any isolation.

To me the most reasonable way to implement seems to be similar to what we do in VS / dotnet test. A remote client similar to vstestconsolewrapper, that talks with processes over some transport, automating the from the outside, running in separate processes. The speed limitation here is startup of the process and engine, and side by side builds of the different mutations. But the isolation is the best between tests.

> Before long we are rebuilding vstest 😅

I don't think that is a bad thing, you'd be re-building the client we have in VS  and dotnet test (for MTP those are 2 different clients), and I don't think there is that much work that needs to be done (but I am not stryker expert :) )

**Comment by slang25** [comment]
Thanks folks, have updated the requirements.

I think my mental model of MTP might be a bit off, I thought we could execute the test project (which is self-executing) and then talk to it via gRPC to tell it what tests to run. If this were true (and I suspect it's not 😅) then we could spin up the processes ahead of time, pooling them, and then grab a running process and start running the tests.

**Comment by dupdob** [comment]
I am afraid that the overall testing infrastructure gets more and more hostile to Stryker-like use cases. I.e, it optimizes for fast and safe execution of full test base vs the ability to fine control test runs.
One of the first victims of this trend is the actual test isolation. Test isolation is a core tenet of good unit testing and fundamental for Stryker's reliability.  Alas, test case generator patterns break isolation. Plus test frameworks do not report this test case accurately.

In an ideal world, Stryker just needs to be able to run any tests list of this choosing and be able to abort it after a first failure (optional). If tests are properly isolated, this can happen within the confine of a single process.

Because test process setup/tear down is costly (around half a second with VsTest), Stryker needs to use as few as possible processes.

Note that thanks to coverage analysis, Stryker is able to test more than one mutation simultaneously, assuming any test only covers one of the active mutations.

Any way, I feel that the best way to integrate MTP with Stryker implies instrumenting the test projects so they behave as needed. But, may be this instrumentation does not need to be automatically injected by Stryker but could be purposely added by the user.

Ideally, he/she would just need to add the proper Stryker package to his/her test project and the this package will provide the needed logic.

**Comment by nohwnd** [comment]
>  If tests are properly isolated,

What isolation do you mean here exactly?

I am talking about running the test in an appdomain that will re-initialize statics etc. We don't have that in .NET anymore, same as thread aborts, because process level isolation is preferred, and so we have to work with that in MTP as well. When the runtime does not offer isolation, we cannot offer isolation.

But maybe we are each talking about a different thing.

If this level of isolation is not required then we can work within the limits of the process, and do basically anything, we don't want to be hostile to stryker, the opposite actually :)

**Comment by slang25** [comment]
I believe Microsoft uses stryker internally, so certainly wants to see it succeed here.

**Comment by rouke-broersma** [comment]
> The speed limitation here is startup of the process and engine, and side by side builds of the different mutations.

This is one thing we were hoping would be improved with MTP since we can have more direct control over execution and shutdown. We may still be able to achieve this, since we know when mutants are in a static context. We can offer different levels of isolation with the least isolated method reusing the testrunner for all mutants except those that have to be isolated. If we can do that, then we can also offer a way to run all mutations in isolation (similar to modes we have now, but then with more direct control).

> I think my mental model of MTP might be a bit off, I thought we could execute the test project (which is self-executing) and then talk to it via gRPC to tell it what tests to run. If this were true (and I suspect it's not 😅) then we could spin up the processes ahead of time, pooling them, and then grab a running process and start running the tests.

We can do that, if we inject the communication code and also the code that re-uses the test project. MTP itself is entirely self executing so it doesn't have anything for this, but it does contain extension points where we could inject our own communication and test ordering logic. That is something @richardwerkman is already working on and seems to be entirely possible. The hard part is that we will have to calculate all the different combinations of testrunner, targetframework, source project etc so we know which runners we need to spin up, when they can be reused. Basically what vstest currently does.

**Comment by dupdob** [comment]
> > If tests are properly isolated,
>
> What isolation do you mean here exactly?

Properly isolated tests should not depend on any (shared) static. I understand this is an ideal situation which is rarely achieved.

**Comment by slang25** [comment]
I guess it's a requirement for mixed-mutant test runs, which I've never had a problem with.

**Comment by nohwnd** [comment]
> Properly isolated tests should not depend on any (shared) static. I understand this is an ideal situation which is rarely achieved. Furthermore, tests must also be isolated when they fail, which is even harder to achieve.

I do agree with you here. But we cannot give any guarantees. Which was how I approached your comment. 

If user does the work themselves, and keeps good practice to make tests independent, not influencing each other in any way even if they fail. Then I don't see a problem with running them in one process and scheduling them in a way that is similar to how hotreload extension does it.

But again I am not an expert in stryker-dotnet. @richardwerkman I assume the code you wrote is internal, and not public, and we would need to have a private call with you and your team to look at it?

**Comment by rouke-broersma** [comment]
> > Properly isolated tests should not depend on any (shared) static. I understand this is an ideal situation which is rarely achieved. Furthermore, tests must also be isolated when they fail, which is even harder to achieve.
> 
> I do agree with you here. But we cannot give any guarantees. Which was how I approached your comment.
> 
> If user does the work themselves, and keeps good practice to make tests independent, not influencing each other in any way even if they fail. Then I don't see a problem with running them in one process and scheduling them in a way that is similar to how hotreload extension does it.
> 
> But again I am not an expert in stryker-dotnet. [@richardwerkman](https://github.com/richardwerkman) I assume the code you wrote is internal, and not public, and we would need to have a private call with you and your team to look at it?

So far it's been experiments so I don't think Richard pushed the code yet but our work on stryker is fully open source so we can publish it, no problem. @richardwerkman are you experiments in a state that it would be useful to share with @nohwnd ?

**Comment by richardwerkman** [comment]
@nohwnd I've prepared a little sample on a branch. https://github.com/stryker-mutator/stryker-net/blob/f6556472dbb4cc0d83d88ad9c832b0b6843594a8/MTP%20Demo/MsTestRunnerDemo/MsTestRunnerDemo/Program.cs#L6

I've kinda hit a dead end here I think with the injected code method. It works, as long as the target framework of the executing assembly is the same as in the test project. But then the whole "bennefit" of the code injection is gone... So I see two solutions: running in a new process with the dotnet cli. But this is slow and gives less control. Or making sure the correct runtimes are loaded into stryker. But that seems like a no go to me.

Do you thinks there is another solution for this? Maybe there is another way of starting MTP that I haven't found yet.

Also I'm curious if you have a sample from VS or dotnet test where MTP is used with the hotreload plugin? That looks promising. So far I could only rerun tests by calling `TestApplication.BuildAsync()` multiple times and then running the build test application.

**Comment by nohwnd** [comment]
Looking at the sample code it will load a dll into an executable, and run it's main method multiple times. This should work just fine a long as the executing assembly is targeting the same target framework **family** as the loaded dll, meaning .NET process needs to load .NET, and .NET Framework needs to load .NET Framework. Of course architectures also need to agree, but I am assuming AnyCPU.

The process needs to use appropriate target framework version as well, matching (or higher) than the highest TFM of the loaded assemblies.

But overall I don't see why this should not work. This is pretty much how testhost.exe works in vstest, even though there we build it with the newest SDK, targeting the lowest supported TFM and then provide deps.json and runtime.config.json that is appropriate for the version of the tested assembly, but if we always provided net10 all dlls would load just fine. We also don't re-use testhost because we are afraid of cross-pollution, but we could.

And here I reach the point where I am missing the details of how stryker works :) I am assuming what you are trying to achieve here is that you start 1 process, load an assembly into it, and run it's tests. Then you mutate? And then you  run again? 

Or do you have a lineup of mutated dlls that you can run, and by the env variables that you set you are able to reach the different mutations without unloading the dll?


For hot reload, I don't think there is any public implementation of the hot-reload part that is done on VS side, the test client on VS side detects that hot reloading happened, and tells the dll to run tests again. Details of how that happens in the assembly can be found here: https://github.com/microsoft/testfx/blob/main/src/Platform/Microsoft.Testing.Extensions.HotReload/HotReloadTestHostTestFrameworkInvoker.cs 

But I don't know if you find details of how the assembly is actually patched, I've asked around if this is tied to VS or also available for others to use.

Demo of how it is used can be seen here, I have no headphones here, but hopefully I am linking close enough to the start of the demo: https://youtu.be/9Jz47Ze9LOI?si=AymLqdCBZapWV1n7&t=1243

**Comment by rouke-broersma** [comment]
> Or do you have a lineup of mutated dlls that you can run, and by the env variables that you set you are able to reach the different mutations without unloading the dll?

Yes, this is what we do in stryker. We mutate once, place all mutations in branches and activate mutations as needed. We then run the unit tests that are relevant for the enabled mutation to limit runtime and we collect the result te determine if mutants are killed/survived.

**Comment by nohwnd** [comment]
In that case it seems to me like it should work with the approach that is similar to hot reload extension (or maybe even that one). Personally I would stay with the constraints of the same test project (but maybe not necessarily the same tfm, just within the same family). Because then the project has the correct deps.json.

Too bad we don't have an official client yet, that would make doing a protype for you much easier. Will try to whip something up and talk with people in our team.

**Comment by richardwerkman** [comment]
> But overall I don't see why this should not work. This is pretty much how testhost.exe works in vstest, even though there we build it with the newest SDK, targeting the lowest supported TFM and then provide deps.json and runtime.config.json that is appropriate for the version of the tested assembly, but if we always provided net10 all dlls would load just fine. We also don't re-use testhost because we are afraid of cross-pollution, but we could.

So that would mean we (as stryker) have to stay up to date with the latest framework all the time? Currently we only update for LTS releases. But even then that would mean we can't run testprojects that use a preview version.

**Comment by rouke-broersma** [comment]
> > But overall I don't see why this should not work. This is pretty much how testhost.exe works in vstest, even though there we build it with the newest SDK, targeting the lowest supported TFM and then provide deps.json and runtime.config.json that is appropriate for the version of the tested assembly, but if we always provided net10 all dlls would load just fine. We also don't re-use testhost because we are afraid of cross-pollution, but we could.
> 
> So that would mean we (as stryker) have to stay up to date with the latest framework all the time? Currently we only update for LTS releases. But even then that would mean we can't run testprojects that use a preview version.

Only the test host would need to be up-to-date, right?

**Comment by nohwnd** [comment]
The test project itself would be your host, you would be just automating it from the outside via an orchestrator.  Similarly to what dotnet test does for example.

**Comment by nohwnd** [comment]
Posted too soon :) 

stryker.exe -> testProject1.exe, all test projects need to install your MTP based extension as nuget, that is similar to what our hotReload extension does. Stryker.exe (or some other global tool, that I think you already have) will orchestrate the testproject1.exe and other processes. And will tell them:

load testproject1.dll, and run with those env variables to run mutations,
done? 
okay now run testproject1.dll with these other env variables to run other mutations

as Richard shown above in the prototype.

**Comment by nohwnd** [comment]
Stryker.exe can target any tfm, its purpose would be to mutate and build the different versions of the mutated projects, by calling dotnet build or however you do it. As long as that dotnet build comes from dotnet SDK that supports net10 preview then you can build net10 projects, and because there is air gap between your stryker.exe orachestrator and the testProject1.exe you don't care about the TFM of testProject1.exe.

**Comment by nohwnd** [comment]
https://github.com/microsoft/testfx/blob/1b5657e065cb412244f2e677dcd30db6eda6b8b2/docs/mstest-runner-protocol/001-protocol-intro.md?plain=1 talking would be done over this protocol, for which we don't have official client package yet. 

But the spec is open, and tunit implements the messages you'd need:  https://github.com/thomhurst/TUnit/blob/main/TUnit.RpcTests/Clients/TestingPlatformClient.cs

**Comment by nohwnd** [comment]
This is how you start the exe in server mode https://github.com/thomhurst/TUnit/blob/main/TUnit.RpcTests/Tests.cs#L40

Similar stuff in our own server mode tests, https://github.com/microsoft/testfx/blob/main/test/UnitTests/Microsoft.Testing.Platform.UnitTests/ServerMode/ServerTests.cs

**Comment by rouke-broersma** [comment]
How would that work with multiple test assemblies? Orchestrate multiple test hosts (test projects) for the same env and aggregate the results ourselves?

**Comment by nohwnd** [comment]
> How would that work with multiple test assemblies? Orchestrate multiple test hosts (test projects) for the same env and aggregate the results ourselves? Because before I was thinking we would create a generic test host that implements MTP -> load all needed test project dll's instead of letting the test project be self-executing.

Yes like that. You can make a project that aggregates multiple dlls test dlls. I think we even natively support it in mstest, because you can provide multiple test dlls to the mstest invocation (addMSTest takes an array, https://grep.app/search?q=addmstest) I am not sure if this path is well tested but should be possible.

Personally I would look into the 1 testhost per test project approach, because that won't force you to "mix" dependencies of test projects, and the tested code, so you are actually testing what you deploy, and it also is easier to run in parallel.

> auto instrument

Yeah that is a good idea. Should work either way.

So do you think the orchestration would work for Styker? So I don't waste much time making prototype for something you will not be able to use.

**Comment by rouke-broersma** [comment]
I think the orchestration would work fine. This is how we currently broadly speaking orchestrate VsTest:


![Image](https://github.com/user-attachments/assets/8599175b-8be9-4068-9d0f-1cfda2c3f9b9)


I will try to draw how I imagine it would work with MTP in a minute, you can let me know if that makes sense

**Comment by nohwnd** [comment]
Tagging myself @nohwnd  to go back to this on monday.

**Comment by slang25** [comment]
@nohwnd I've had some time over the weekend to do some research here and hack around.

In VSTest we have the idea of `DataCollection` for out of process data collection. Is there an equivalent in MTP?

**Comment by 0xced** [comment]
Paging @Youssef1313 who will certainly be able to offer good advice on this topic.

**Comment by Youssef1313** [comment]
Thanks for the ping @0xced.

I haven't read the full thread, but on the last question for out-of-process extensibility, we have it in MTP.

We have a sample for that: https://github.com/microsoft/testfx/tree/605a2c741c420b1b7a6877fcce5a0fd69bb80e53/samples/public/TestingPlatformExamples/TestingPlatformExplorer/Out-of-process%20extensions

The 2 most important type of hosts that we have in MTP:

1. TestHost: This is the process where the tests actually run
2. TestHostController: This is our out-of-process extension point, common for stuff like code coverage and crash/hang dump extensions.

**Comment by Lorilatschki** [comment]
Any progress on that?

**Comment by rouke-broersma** [comment]
We finally have a plan and a workable direction! But progress is slow.

**Comment by thomhurst** [comment]
> We finally have a plan and a workable direction! But progress is slow.

TUnit guy here - will the planned approach just work natively for all MTP test runners? No work or custom behaviour needed for the individual test frameworks?

**Comment by rouke-broersma** [comment]
It should not require anything on the test framework side, but may require test framework specific processing on our side just as we have for vstest currently. Mostly because some test frameworks have features that don't follow spec.

**Comment by thomhurst** [comment]
Let me know if you need anything from me!

**Comment by Lorilatschki** [comment]
Since we have completely switched to MTP 2.0.0, we are not able to use stryker unless MTP support is built in. Is there at least a rough plan for when support might be available?

**Comment by richardwerkman** [comment]
@Lorilatschki  I am still working on MTP support, but progress has been slow. I have a branch `mtp_runner` which is getting to a point where I can run MTP tests and see mutants getting killed. 

Outstanding issues:
- At the end of the run, the last few mutants are not being picked up by the testrunner. This causes stryker to wait indefinitely..
- Coverage analysis hasn't been implemented yet
- I've only tested with MSTest so far, there might be more issues with other test frameworks or configurations

The good news:
Once I have something that is at least finishing a run and giving usefull output we can release as a POC. Since the testrunner has to be enabled by using the config `--testrunner mtp` we can safely release without breaking changes. From there we can iterate further to implement further fixes and features like coverage analysis

**Comment by richardwerkman** [comment]
A status update: I've made great progress. I've done basic testing on all MTP test frameworks now (xUnitv3, NUnit4, MSTestv3, TUnit). Just some cleaning up to do and we can release a preview version. I expect to be able to release this week. 

I can use some help with testing. We will need to run stryker on projects both with the VSTest and MTP testrunner and compare the results. I'm suspecting there could be issues with static mutants since they are handled differently in MTP.

After that I plan to continue working on this. We will need to add support for coverage analysis. I'm suggesting keeping this issue open to keep track of testing the preview version and coverage analysis support.

**Comment by richardwerkman** [comment]
PR #3404 has been merged. I'm releasing stryker v4.12.0 now which includes the `--testrunner` flag. It can be set to `mtp` manually for now to use the new runner. It can also be enabled from `stryker-config` file. I'll keep working on it, with a focus on adding coverage analysis. Any feedback would be appreciated!

**Comment by Lorilatschki** [comment]
@richardwerkman I tried to update to v4.12.0 but it is not yet available at nuget.org. Is it because you are still in the release process or do you deploy it to a nuget repo other than nuget.org?

**Comment by rouke-broersma** [comment]
@Lorilatschki check again :)

**Comment by Lorilatschki** [comment]
@richardwerkman in one of our test projects, which are covering the stryker usecases for our central provided ci/cd gitlab pipelines, it looks promising. The tests can be executed via `--testrunner mtp` 👍 

The warning `The Microsoft Test Platform testrunner is currently in preview. Coverage analysis is currently unsupported and results should be verified since this feature is still being tested.` is very usefull, looking forward to get this feature 😃 

Many thanks for your great support! You are awesome! 🚀

**Comment by rouke-broersma** [comment]
@Lorilatschki please also let us know any results you're seeing such as execution speed, differences in results etc. Missing coverage analysis mostly means that we are missing an optimalization and that we are not able to give you the same detail level in reports, but should not have a a negative impact on accuracy (actually running without coverage analysis is the most accurate since we don't skip any unit tests). 

There are accuracy problems for other reasons, for example we don't yet properly support static context so if you use static context in your source code your mutation result may be drastically wrong in some places.

**Comment by lushess255** [comment]
I tried mutating .NET Framework 4.7.2 MSTest unit tests but got various "Stryker failed to test 1 mutant(s)." errors.  This is my first time doing this, so I'm not sure if the errors are due to my mistakes or the new MTP support.  I'm running from the source directory, referencing the solution and test-projects in stryker-config.json plus I included "coverage-analysis": "off" as well.

The command line I used was: dotnet stryker --testrunner mtp -o

[stryker_log.txt](https://github.com/user-attachments/files/25024474/stryker_log.txt)

**Comment by rouke-broersma** [comment]
@lushess255 I don't think dotnet framework is supported in MTP, so we also cannot support it. You should stick with vstest. @richardwerkman we should probably build a guard for this.

**Comment by thomhurst** [comment]
MTP does work with framework. Tunit only uses MTP and you can create a net framework test project with it

**Comment by Lorilatschki** [comment]
When running stryker for our productive repos (we couldn't run it for a while, since we have switched to MTP), we are getting the following error. (I have tested it for 2 different repos, which I cannot share due to data reasons).

This is the command line:
```shell
dotnet-stryker --output .stryker/run_1 --project Add.Cli.csproj --test-project test/Add.Cli.Test/Add.Cli.Test.csproj --test-project test/StWriter.Test/StWriter.Test.csproj --test-project test/StWriter.Test.Unit/StWriter.Test.Unit.csproj --test-project test/Cli.Test.Unit/Cli.Test.Unit.csproj --test-project test/Cli.Test/Cli.Test.csproj --test-project test/FileIo.Test/FileIo.Test.csproj --test-project test/FileIo.Test.Unit/FileIo.Test.Unit.csproj --test-project test/Add.Cli.Test.Unit/Add.Cli.Test.Unit.csproj  --testrunner mtp --mutation-level Advanced
```

This is the output:
```
Version: 4.12.0
[05:50:37 INF] Analysis starting.
[05:50:37 INF] Stryker will mutate solution SafetyCli.
[05:50:37 INF] Using configuration/platform 'Debug|Any CPU'.
[05:50:37 INF] Identifying projects to mutate in /SafetyCli.sln. This can take a while.
[05:50:45 INF] Could not find an assembly reference to a mutable assembly for project /test/Cli.Test.Unit/Cli.Test.Unit.csproj. Will look into project references.
[05:50:45 INF] Could not find an assembly reference to a mutable assembly for project /test/StWriter.Test/StWriter.Test.csproj. Will look into project references.
[05:50:45 INF] Could not find an assembly reference to a mutable assembly for project /test/Cli.Test/Cli.Test.csproj. Will look into project references.
[05:50:45 INF] Could not find an assembly reference to a mutable assembly for project /test/FileIo.Test/FileIo.Test.csproj. Will look into project references.
[05:50:45 INF] Could not find an assembly reference to a mutable assembly for project /test/StWriter.Test.Unit/StWriter.Test.Unit.csproj. Will look into project references.
[05:50:45 INF] Could not find an assembly reference to a mutable assembly for project /test/Add.Cli.Test/Add.Cli.Test.csproj. Will look into project references.
[05:50:45 INF] Could not find an assembly reference to a mutable assembly for project /test/Cli.Test.Apax/Cli.Test.Apax.csproj. Will look into project references.
[05:50:45 INF] Could not find an assembly reference to a mutable assembly for project /test/FileIo.Test.Unit/FileIo.Test.Unit.csproj. Will look into project references.
[05:50:45 INF] Project /src/Add.Cli/Add.Cli.csproj analysis failed hence can't be mutated.
[05:50:45 INF]   referenced by test project /test/Cli.Test.Unit/Cli.Test.Unit.csproj, analysis failed.
[05:50:45 INF]   referenced by test project /test/Add.Cli.Test/Add.Cli.Test.csproj, analysis failed.
[05:50:45 WRN]   can't be mutated because all referencing test projects' analysis failed.
[05:50:45 INF] Test project /test/StWriter.Test/StWriter.Test.csproj does not appear to test any mutable project, analysis failed.
[05:50:45 INF] Test project /test/Cli.Test/Cli.Test.csproj does not appear to test any mutable project, analysis failed.
[05:50:45 INF] Test project /test/FileIo.Test/FileIo.Test.csproj does not appear to test any mutable project, analysis failed.
[05:50:45 INF] Test project /test/StWriter.Test.Unit/StWriter.Test.Unit.csproj does not appear to test any mutable project, analysis failed.
[05:50:45 INF] Test project /test/Cli.Test.Apax/Cli.Test.Apax.csproj does not appear to test any mutable project, analysis failed.
[05:50:45 INF] Test project /test/FileIo.Test.Unit/FileIo.Test.Unit.csproj does not appear to test any mutable project, analysis failed.
[05:50:45 WRN] Use --diag option to have the analysis logs in the log file.
[05:50:45 INF] Analysis complete.
[05:50:45 INF] Time Elapsed 00:00:08.2930823
Stryker.NET failed to mutate your project. For more information see the logs 
below:
Failed to analyze project builds. Stryker cannot continue.
```

The diag log can be found as attachment and it took really long ~3:30m. _I have redacted all sensitive data._

[output.log](https://github.com/user-attachments/files/25044649/output.log)

**Comment by richardwerkman** [comment]
@lushess255 I didn't test for .net Framework yet since I work on macOS. Good find. We should create an integration test for that scenario and fix the issue.

**Comment by richardwerkman** [comment]
@Lorilatschki I'm not sure if this is related to MTP.. It seems more related to mutating the assembly. It might be related to how stryker is triggered. Have you tried running from solution context? Then stryker loads all projects from the solution file and you don't have to supply the `--test-project` options.

**Comment by rouke-broersma** [comment]
> [@Lorilatschki](https://github.com/Lorilatschki) I'm not sure if this is related to MTP.. It seems more related mutating the assembly. It might be related to how stryker is triggered. Have you tried running from solution context? Then stryker loads all projects from the solution file and you don't have to supply the `--test-project` options.

This seems related to the improved project resolution and solution mode support

**Comment by Lorilatschki** [comment]
> [@Lorilatschki](https://github.com/Lorilatschki) I'm not sure if this is related to MTP.. It seems more related mutating the assembly. It might be related to how stryker is triggered. Have you tried running from solution context? Then stryker loads all projects from the solution file and you don't have to supply the `--test-project` options.

No I haven't tried that, since we should not use the `--solution` parameter since this is only required for net framework projects. See [issue 3399](https://github.com/stryker-mutator/stryker-net/issues/3399#issuecomment-3732542654) for reference.

**Comment by rouke-broersma** [comment]
@Lorilatschki you are using solution mode: `Identifying projects to mutate in /SafetyCli.sln` is found in your logs. Also I think the mentioned issue should be fixed now.

**Comment by dupdob** [comment]
I am looking into this.
@Lorilatschki: I don't know if the feature does not work as intended or if its a consequence of the redaction of the logs, but I was expecting the build logs (in diag mode) but alas non shows.
The build gets slower because diag mode disables concurrency and improves on default retry logic in an attempt to overcome whatever is failing.

**Comment by richardwerkman** [comment]
@Lorilatschki Stryker will run in solution context when it detects a solution file in the directory it is invoked in. You don't have to pass the `--solution` option to enable this context. In your logs I see: `"IsSolutionContext": true`, so I guess that is the case for you. Maybe your issue is related to somehow running both in solution context and target project context. Target project context is enabled when the `--test-project` option is used. 

I would try running in non solution context with the same options (run from a different directory, the dir of your target project). Or remove the `--test-project` options to only use the solution context.

**Comment by Lorilatschki** [comment]
> [@Lorilatschki](https://github.com/Lorilatschki) Stryker will run in solution context when it detects a solution file in the directory it is invoked in. You don't have to pass the `--solution` option to enable this context. In your logs I see: `"IsSolutionContext": true`, so I guess that is the case for you. Maybe your issue is related to somehow running both in solution context and target project context. Target project context is enabled when the `--test-project` option is used.
> 
> I would try running in non solution context with the same options (run from a different directory, the dir of your target project). Or remove the `--test-project` options to only use the solution context.

Thanks for the hint. Unfortunatelly the solution often contains Integration tests, so I try to change the directory.

**Comment by richardwerkman** [comment]
You could look into using the `test-case-filter` option to exclude the integration tests when running in solution context.

**Comment by Lorilatschki** [comment]
> You could look into using the `test-case-filter` option to exclude the integration tests when running in solution context.

Will check it. However, for me it would be more intuitive to ignore the solution context completly, once a project context is specified.

**Comment by richardwerkman** [comment]
Status update: I've made progress on implementing coverage analysis for MTP. It now filters uncovered mutants, but doesn't yet filter uncovered testcases per mutant. This has been merged to main, however there are still some blocking issues before we can release a new version. I think the next version will still be a preview version for the MTP runner. After that we can start to think about a stable version.

I've renamed the option from `--testrunner` to `--test-runner` because all other options are using kebab case as well. This is a breaking change but I didn't do backwards compatibility because we're still in preview for this feature.

I hope to be able to release a new preview version soon.

**Comment by thomhurst** [comment]
Would this be helpful? https://github.com/microsoft/testfx/issues/5554

**Comment by richardwerkman** [comment]
I've merged #3460 which contained a few important fixes. Will create a release today. 

That should make the mtp runner pretty usable, since it can handle timeouts correctly and has coverage analysis implemented.

I'm thinking to still keep the feature in preview since we haven't tested on many real world projects yet.

Still todo:

- [x] Add test reporting for html and json report
- [ ] Verify .net framework support
- [ ] Do per test timeout calculation

**Comment by iamwyza** [comment]
Is it known that the "test-projects" config is seemingly ignored when running in MTP?

**Comment by richardwerkman** [comment]
Status update: I found a breaking issue in the runner, related to StackOverflowExceptions. After I fixed that I was able to run stryker both with VSTest and MTP on the same project (regex-parser recently migrated so I just had to rollback a few commits).

VSTest:
<img width="1276" height="241" alt="Image" src="https://github.com/user-attachments/assets/3fb4513d-d0c7-4b2b-b338-4cabfb570ac0" />

MTP:
<img width="1301" height="240" alt="Image" src="https://github.com/user-attachments/assets/afd1727d-3a01-4ecc-aba1-ca488ed9a2e8" />

The results are interesting! The mutantionscore is correct on MTP (1 mutant gets status RuntimeError, which we didn't implement correctly in VSTest). That is a nice validation that the runner works like expected. However, VSTest is over 3 times faster! Not what I expected at all. I'll investigate why MTP is so slow.

**Comment by richardwerkman** [comment]
The `StackOverflowException` bug has been fixed. The performance hit has also been fixed. It was a problem with inter-process communications. @rouke-broersma can we release a new version with these fixes?

**Comment by 0xced** [comment]
Congrats on the 4.15.0 release, the MTP runner is working fine on my [xunit-v3-mtp](https://github.com/serilog-contrib/serilog-formatting-log4net/tree/xunit-v3-mtp) branch now!

**Comment by martincostello** [comment]
Just tried the latest release out with Polly here: https://github.com/App-vNext/Polly/pull/3131

Things are definitely improving, but 2/5 of our projects are still failing to work correctly.

**Comment by richardwerkman** [comment]
@martincostello what kind of issues are you seeing? Please open an issue if it is something not on our backlog at the moment. At the moment I'm only aware of the .net framework issue.

**Comment by martincostello** [comment]
There's some details in the linked PR description.

TL; DR: it prints a load of E characters for each test, then says it can't determine the mutation coverage, then exits "successfully", masking the fact it didn't actually meet the required mutation level of 100%. I haven't dug any further.

**Comment by slang25** [comment]
Sorry for the big wall of text, but my agent says this...

---
I've reproduced, root-caused, and fixed the issue. Here's the full picture.

## What was happening

Running Stryker with `--test-runner mtp` against Martin Costello's xunit v3 Polly branch (PR #3131): every mutant printed `E`, then *"Stryker was unable to calculate a mutation score"*, then exited "successfully" — masking that nothing was actually tested.

I reproduced it locally against the real `Polly.Core` (scoping mutations to one file so it ran in ~25s). A minimal xunit v3 + MTP project did **not** reproduce it — the trigger was something specific to Polly.

## Root cause

`E` = `MutantStatus.RuntimeError`, which Stryker assigns when the MTP test host **crashes** during a run. Capturing the host's own output (`--log-to-file`) revealed an unhandled exception killing the process on every run:

```
Could not load file or assembly 'Polly.Core, Version=8.0.0.0, ... PublicKeyToken=c8a3ffc3f8f825cc'
   at System.Reflection.RuntimeAssembly.GetExportedTypes()
   at Xunit.v3.XunitTestFrameworkDiscoverer.GetExportedTypes()
```

Comparing assembly identities:

| Assembly | Version | PublicKeyToken |
|---|---|---|
| Original `Polly.Core.dll` | **8.0.0.0** | c8a3ffc3f8f825cc |
| Stryker's injected mutant | **1.0.0.0** | c8a3ffc3f8f825cc |

Stryker preserves the strong-name *key* but **resets the assembly version to 1.0.0.0**. Polly is strong-named and uses **MinVer** to compute its version (8.x). MinVer only sets `AssemblyVersion` during a *real* build — but Stryker recompiles the mutated assembly from source trees gathered via Buildalyzer's *design-time* build, where MinVer never runs, so the version defaults to 1.0.0.0. For a strong-named assembly the loader is version-strict: the unchanged test assembly still references `Polly.Core 8.0.0.0`, can't bind the downgraded 1.0.0.0 mutant, throws on a background thread, and the whole test host dies before any test runs. My minimal Calc repro isn't strong-named, so version is ignored — which is why it worked.

## The fix

`src/Stryker.Core/Stryker.Core/Compiling/CsharpCompilingProcess.cs` — when building the mutated compilation, read the **original built assembly's version** (`AssemblyName.GetAssemblyName(...)`, metadata-only) and force the mutated compilation to emit that same `AssemblyVersion` (stripping any stale attribute first to avoid a duplicate). Safe no-op when the original can't be read or already matches, so non-strong-named / non-MinVer projects are unaffected. This is general — it fixes any strong-named project using MinVer/Nerdbank.GitVersioning/etc., not just Polly.

## Verification

- **Before:** every mutant `E` → "unable to calculate a mutation score".
- **After (real Polly.Core via MTP):** `Coverage capture complete: 1324 mutations covered`, mutants now report real results `.....................SS.S.S..S..S`, **final score 81.82%**, zero `E`s.
- Non-strong-named Calc repro: unchanged (85.71%, no regression).
- Added 2 unit tests; all 12 compiling + 41 compiling/mutation tests pass.

Changes are staged in the working tree on branch `slang25/taipei` (not committed). Want me to commit and open a PR against `master`? If so I'd also suggest filing the upstream issue Richard asked Martin for, since this is a concrete, reproducible bug with a fix.

One related observation worth flagging separately: even when the host crashed, Stryker exited with success code 0 while reporting a NaN score — arguably it should surface a non-zero exit when it couldn't test any mutants. That's a separate hardening change I can tackle if you want.

**Comment by slang25** [comment]
If you re-run setting MinVerSkip=true, it _should_ work

**Comment by martincostello** [comment]
> If you re-run setting MinVerSkip=true, it _should_ work

Yep, that works.

**Comment by thomhurst** [comment]
#lazy here - but anyone let me know if there's any TUnit shortcomings please if you're using it!

**Comment by martincostello** [comment]
Looks like the only remaining issue for Polly (other than some new mutants I need to add) after applying workaround is the performance.

On main our slowest job takes ~8m. The PR to move to MTP and xunit v3 times out after an hour.

<img width="191" height="180" alt="Image" src="https://github.com/user-attachments/assets/197161bc-769b-46ac-8908-723987bebc1f" />

<img width="199" height="180" alt="Image" src="https://github.com/user-attachments/assets/44e8d2e7-8994-48ac-91b5-72fc30bf1a1b" />

**Comment by rouke-broersma** [comment]
Interesting. The other projects have similar runtime between vstest and mtp so probably still something broken for that legacy project.

**Comment by martincostello** [comment]
I increased the timeout to 2 hours, so hopefully some time today there'll be a full setup of logs for inspection.

The `core` project is nearly 2x slower though, so might not just be isolated to that 1 project.

**Comment by martincostello** [comment]
> I increased the timeout to 2 hours, so hopefully some time today there'll be a full setup of logs for inspection.

[Workflow logs](https://github.com/App-vNext/Polly/actions/runs/28314815126/job/83885970688#step:5:141)

1 hour 38 minutes for 100.00% mutant coverage, and there don't appear to be any errors (the progress is all `.`, no `E`s). Seems like it's just genuinely slower under MTP.

**Comment by richardwerkman** [comment]
> If you re-run setting MinVerSkip=true, it should work

Should we set this by default in Stryker then? It looks like it doesn't hurt any other project.

**Comment by slang25** [comment]
I'd suggest removing the design-time build parameter, which needs some consideration. And it would also be good to set a `STRYKER_BUILD` parameter too, so that people can also react.

I think Stryker already disabled verify interactive diffs and something else, so adding more well known external params cant hurt.

**Comment by richardwerkman** [comment]
About the performance on polly. I think the main issue is that per-test-coverage hasn't been implemented yet. See #3516 for the status. Since polly has quite a large testset with near 100% coverage, this means that basically every test is ran for every mutant that has any coverage, which is exponentially slower on large projects. After #3516 has been merged Stryker should only run the covering tests for each mutant. Since vstest already implements this, it would be a lot faster on a project like polly.

**Comment by richardwerkman** [comment]
I've also started work on implementing bail for MTP #3655. That should also give significant performance boost on large projects. Especially while per-test coverage hasn't been implemented yet.

**Comment by 0xced** [comment]
> The `core` project is nearly 2x slower though, so might not just be isolated to that 1 project.

That's crazy. For me, it's the opposite. The MTP runner is roughly 2x _faster_ than the VSTest runner!

On my machine: 9s with MTP vs 19s with VSTest. On GitHub actions: 34s vs 1m14.

But indeed it's not the same size. My project has only 2k lines of C# code and 177 mutants.

**Comment by dupdob** [comment]
@0xced : VsTest has a significant fixed overhead, .2-.5 seconds per session. This will dominates the run time on small projects and/or really fast tests. MTP uses a far more streamlined architecture, so it runs instantaneously.

**Comment by richardwerkman** [comment]
@martincostello I've got per-test analysis working on a branch. The runtime is now a lot closer to what you expected:

<img width="851" height="149" alt="Image" src="https://github.com/user-attachments/assets/2efb2b05-3e4c-45b8-8df4-74e1d1920fff" />

We can still improve further with bail, I expect to see similar performance to vstest then. But I noticed the mutation score is not 100% like in your previous run. Could you elaborate which mutation score is to be expected for this project? It could be a defect on my end.

**Comment by martincostello** [comment]
@richardwerkman Currently each of our projects has a mutation score of 100%.

**Comment by richardwerkman** [comment]
@martincostello I've verified the mutation score I get on polly with per-test analysis and it seems correct at 90%. The surviving mutants are not detected by the tests when I check manually. So I guess there is another reason why the mutation score is lower on that branch than you expect.

I will open a PR for per-test analysis, we can discuss this further in that PR

**Comment by martincostello** [comment]
Updating Polly to .NET 11 preview 7 is causing all our mutation tests to fail completely ([example](https://github.com/App-vNext/Polly/actions/runs/31515824192/job/93860576646#step:5:112)). I'm guessing something's changed to prioritise MTP further in .NET 11 (unless it's a bug in the SDK).

```text
[17:06:51 INF] Found project /home/runner/work/Polly/Polly/src/Polly.Core/Polly.Core.csproj to mutate.
[17:06:51 INF] Analysis complete.
[17:06:51 INF] Building test project /home/runner/work/Polly/Polly/test/Polly.Core.Tests/Polly.Core.Tests.csproj (1/1)
[17:06:51 INF] Building project Polly.Core.Tests.csproj using dotnet build Polly.Core.Tests.csproj -c Debug --property:Platform=AnyCPU (directory /home/runner/work/Polly/Polly/test/Polly.Core.Tests.)
[17:07:08 ERR] TestDiscoverer: Test discovery has been aborted!
[17:07:08 WRN] Project '/home/runner/work/Polly/Polly/test/Polly.Core.Tests/Polly.Core.Tests.csproj' did not report any test. This may be because the test adapter package, xunit.runner.visualstudio, failed to deploy or run. Check if any dependency is missing or there is a version conflict, check the testdiscovery logs or explore with VsTest.console.
[17:07:08 INF] Number of tests found: 0 for project /home/runner/work/Polly/Polly/src/Polly.Core/Polly.Core.csproj. Initial test run started.
[17:07:08 WRN] Runner 1: Test assembly does not contain any test, skipping.
```

If that is the case, lack of support for MTP might become a blocker to using stryker with .NET 11 still not using MTP.

**Comment by bradwilson** [comment]
It doesn't appear to be directly .NET 11 Preview 7 causing this, since a non-MTP v3 test project can run using VSTest (unfolded with `dotnet new xunit3 --test-runner vstest` from `4.0.0-pre.154`, then manually updated to target `net11.0`):

<img width="799" src="https://github.com/user-attachments/assets/f845fbd5-f2ab-4c85-a4e4-fc5119fc5d4c" />

**Comment by richardwerkman** [comment]
#3752  Coverage analysis has been merged. It seems to work well, but obviously we will need some real world testing to verify. We can't release a new version yet since we are blocked by some breaking changes and a blocking defect. Hopefully we can release soon 👍

**Comment by martincostello** [comment]
Stryker v5.0.0 has resolved all the issues we were having with xunit.v3 with Polly, and upgrading to xunit.v3 has then resolved the issues with .NET 11 I noted in https://github.com/stryker-mutator/stryker-net/issues/3094#issuecomment-5264697729.
